sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTranspose.scala

```diff
@@ -0,0 +1,148 @@
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.catalyst.trees.TreePattern.TRANSPOSE
+import org.apache.spark.sql.types.{AtomicType, DataType, StringType}
+import org.apache.spark.unsafe.types.UTF8String
+
+
+class ResolveTranspose(sparkSession: SparkSession) extends Rule[LogicalPlan] {
+
+  private def leastCommonType(dataTypes: Seq[DataType]): DataType = {
+    if (dataTypes.isEmpty) {
+      StringType
+    } else {
+      dataTypes.reduce { (dt1, dt2) =>
+        TypeCoercion.findTightestCommonType(dt1, dt2).getOrElse {
+          throw new IllegalArgumentException(s"No common type found for $dt1 and $dt2")
+        }
+      }
+    }
+  }
+
+  private def transposeMatrix(
+    fullCollectedRows: Array[InternalRow],
+    nonIndexColumnNames: Seq[String],
+    nonIndexColumnDataTypes: Seq[DataType]): Array[Array[Any]] = {
+    // Construct the original matrix
+    val originalMatrixNumCols = fullCollectedRows.head.numFields - 1
+    val originalMatrixNumRows = fullCollectedRows.length
+
+    val originalMatrix = Array.ofDim[Any](originalMatrixNumRows, originalMatrixNumCols)
+    for (i <- 0 until originalMatrixNumRows) {
+      val row = fullCollectedRows(i)
+      for (j <- 0 until originalMatrixNumCols) {
+        originalMatrix(i)(j) = row.get(j + 1, nonIndexColumnDataTypes(j))
+      }
+    }
+
+    // Transpose the original matrix
+    val transposedMatrix = Array.ofDim[Any](originalMatrixNumCols, originalMatrixNumRows)
+    for (i <- 0 until originalMatrixNumRows) {
+      for (j <- 0 until originalMatrixNumCols) {
+        transposedMatrix(j)(i) = originalMatrix(i)(j)
+      }
+    }
+
+    // Insert nonIndexColumnNames as first "column"
+    val finalMatrix = Array.ofDim[Any](originalMatrixNumCols, originalMatrixNumRows + 1)
+    for (i <- 0 until originalMatrixNumCols) {
+      finalMatrix(i)(0) = UTF8String.fromString(nonIndexColumnNames(i))
+    }
+    for (i <- 0 until originalMatrixNumCols) {
+      for (j <- 1 until originalMatrixNumRows + 1) {
+        finalMatrix(i)(j) = transposedMatrix(i)(j - 1)
+      }
+    }
+
+    finalMatrix
+  }
+
+  override def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperatorsWithPruning(
+    _.containsPattern(TRANSPOSE)) {
+    case t @ Transpose(indexColumn, child, _, _) if !t.resolved =>
+      if (!indexColumn.dataType.isInstanceOf[AtomicType]) {
+        throw new IllegalArgumentException(
+          s"Index column must be of atomic type, but found: ${indexColumn.dataType}")
+      }
+
+      // Cast the index column to StringType
+      val indexColumnAsString = Alias(
+        Cast(indexColumn, StringType),
+        indexColumn.asInstanceOf[Attribute].name)().asInstanceOf[NamedExpression]
+
+      // Cast non-index columns to the least common type
+      val nonIndexColumnsAttr = child.output.filterNot(
+        _.exprId == indexColumn.asInstanceOf[Attribute].exprId)
```

Comment:
what if index column is not an attribute? such as a nested field like `a.b`

Reply:
I modified to enforce index column to be an attribute (of AtomicType)